Better structure for the internal organization of our state directory

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#564 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have fixed an issue where the parent directories of `--state-dir` wouldn't be created and fail in a not-so-user-friendly way:

```
$ cardano-wallet-http-bridge launch --state-dir /tmp/a/b/c
cardano-wallet-http-bridge: /tmp/a/b/c: createDirectory: does not exist (No such file or directory)
```

- [x] I have now used non-overlapping namespaces for storing chain's data for the different backends

# Comments

<!-- Additional comments or screenshots to attach if any -->

```
$ tree ~/.cardano-wallet/
/home/ktorz/.cardano-wallet/
├── httpbridge
│   ├── mainnet
│   │   ├── mainnet
│   │   │   ├── blob
│   │   │   ├── chainstate
│   │   │   ├── config.yml
│   │   │   ├── epoch
│   │   │   ├── index
│   │   │   ├── pack
│   │   │   ├── refpack
│   │   │   └── tag
│   │   ├── wallet.db
│   │   ├── wallet.db-shm
│   │   └── wallet.db-wal
│   └── testnet
│       ├── testnet
│       │   ├── blob
│       │   ├── chainstate
│       │   ├── config.yml
│       │   ├── epoch
│       │   ├── index
│       │   ├── pack
│       │   ├── refpack
│       │   └── tag
│       ├── wallet.db
│       ├── wallet.db-shm
│       └── wallet.db-wal
└── jormungandr
    └── testnet
        ├── chain
        │   ├── blocks.sqlite
        │   ├── blocks.sqlite-shm
        │   └── blocks.sqlite-wal
        ├── jormungandr-config.json
        ├── wallet.db
        ├── wallet.db-shm
        └── wallet.db-wal
```

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
put state directory in a namespace according to the target + network

KtorZ committed:
recursively create parent directory of state dir when they don't exist

KtorZ committed:
rename jormungandr storage option from 'jormungandr' to 'chain'

Such that our state dir looks nicer:

/home/ktorz/.cardano-wallet/
└── jormungandr
    └── testnet
        ├── chain
        │   ├── blocks.sqlite
        │   ├── blocks.sqlite-shm
        │   └── blocks.sqlite-wal
        ├── jormungandr-config.json
        ├── wallet.db
        ├── wallet.db-shm
        └── wallet.db-wal

KtorZ committed:
re-generate nix machinery

piotr-iohk reviewed (approved):
lgtm :+1: 
(but tests need to be adjusted)

rvl commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1128,14 +1133,25 @@ decodeError bytes = do
 -- | Resolve '~' or '$HOME' in a 'FilePath' to their actual system value
-resolveHomeDir :: FilePath -> IO FilePath
+resolveHomeDir
+    :: forall t. Typeable t
```
This looks like voodoo to me. It might be clearer to split out a more specialised function:
```haskell
configSubDir :: forall n t. (t ~ Jormungandr n, KnownNetwork n) => FilePath
configSubDir = backend </> networkName
  where ...
```

rvl commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1128,14 +1133,25 @@ decodeError bytes = do
     return
         $ T.unpack
         $ T.replace "$HOME" homeDir
         $ T.replace "~" homeDir
```
Had I have seen this before, I would have objected.
1. This substitution should be the job of the command shell.
2. It looks weird on windows.
3. The default of `~/.cardano-wallet` is inconsiderate to users because it clutters their home directory. Better to use `$XDG_DATA_HOME`

The [xdg-basedir](http://hackage.haskell.org/package/xdg-basedir) package will help.

rvl reviewed (dismissed):
I like that parent directories are created and there are by default separate state directories for each network of each backend.

But nesting 2 levels of subdirectories within the state dir will make deployment less convenient. We are also committing the same error as http-bridge does -- see how that causes an unnecessary nesting of directories.

I think we should change it to `stateDirOption :: Parser (Maybe FilePath)`. Adjust the help string to say the default is `$HOME/.cardano-wallet/BACKEND/$network`. Then fill in the default value if the option was Nothing.

Commit:
Adjust tests to the changes

Commit:
Add tests for _nested dir_ case in Jormungandr and http-bridge

rvl committed:
launcher: Use XDG data dir for default --state-dir

- Put the default --state-dir within the XDG data dir
- Make --state-dir point to the actual state dir, rather than nesting
  directories within.

KtorZ committed:
fix code style and re-review test assertions to match new default dir

KtorZ committed:
ignore 'Monoidal Law' in HLint v2.2.1

We do use 'mempty' often to align text and ease readability. So the small overhead is acceptable
and, most probably removed by GHC at compile-time...
